src/main/java/com/conveyal/r5/profile/FrequencyRandomOffsets.java

```diff
@@ -179,10 +224,27 @@ public void randomize () {
                     }
                 }
             }
-
-            if (remainingAfterPreviousRound == remaining && remaining > 0) {
+            if (remainingAfterPreviousPass == remaining && remaining > 0) {
                 throw new IllegalArgumentException("Cannot solve phasing, you may have a circular reference!");
             }
+            // Copy results of randomization to a Map keyed on TripSchedules (instead of TripPattern index ints). This
+            // allows looking up offsets in a context where we only have a filtered list of running frequency trips and
+            // don't know the original unfiltered index of the trip within the pattern. Ideally we'd just build the
+            // original map keyed on TripSchedules (or hypothetical FreqEntries) but that reqires a lot of refactoring.
+            offsetsForTripSchedule.clear();
```

Comment:
What effect does this have? Is it just a safeguard, or is it ever expected that this map will have mappings at this point?

Reply:
Randomize() is called repeatedly, once at each iteration of the raptor search. the first time it's called, the offsets map is indeed empty. But all subsequent times, it contains the offsets from the last time randomize() was called. Clearing an existing map is slightly more efficient than constructing a new one, if only because it already contains an array of exactly the right size to hold the mappings.